src/System.IO.Pipelines.Networking.Tls/SecurePipeline.cs

```diff
@@ -0,0 +1,255 @@
+                                return await _handShakeCompleted.Task;
+                            }
+                        }
+                    }
+                    finally
+                    {
+                        _lowerConnection.Input.Advance(buffer.Start, buffer.End);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Dispose();
+                _handShakeCompleted = new TaskCompletionSource<ApplicationProtocols.ProtocolIds>();
+                _handShakeCompleted.SetException(ex);
+                return await _handShakeCompleted.Task;
+            }
+            finally
+            {
+                StartReading();
+            }
+        }
+
+        private async void StartReading()
+        {
+            try
+            {
+                while (true)
+                {
+                    var result = await _lowerConnection.Input.ReadAsync();
+                    var buffer = result.Buffer;
+                    try
+                    {
+                        if (buffer.IsEmpty && result.IsCompleted)
+                        {
+                            break;
+                        }
+                        ReadableBuffer messageBuffer;
+                        TlsFrameType frameType;
+                        while (TryGetFrameType(ref buffer, out messageBuffer, out frameType))
+                        {
+                            //We have app data or tokens at this point so slice out the message
+                            //If we have app data, we will slice it out and process it
+                            if (frameType == TlsFrameType.AppData)
+                            {
+                                await _contextToDispose.DecryptAsync(messageBuffer, _outputPipeline);
+                            }
+                            else
+                            {
+                                throw new InvalidOperationException("Invalid frame type during the connection was sent");
+                            }
+                        }
+                    }
+                    finally
+                    {
+                        _lowerConnection.Input.Advance(buffer.Start, buffer.End);
+                    }
+                }
+            }
+            finally
+            {
+                try
+                {
+                    //Close down the lower pipeline
+                    _lowerConnection.Input.Complete();
+                    _lowerConnection.Output.Complete();
+                    //Tell the upper consumer that we aren't sending any more data
+                    _outputPipeline.CompleteReader();
+                    _outputPipeline.CompleteWriter();
+                    _inputPipeline.CompleteReader();
+                    _inputPipeline.CompleteWriter();
+                }
+                catch
```

Comment:
Since we're not (necessarily) in an error case, why are we hiding any errors from this?